src/utils/normalize.py

```diff
@@ -0,0 +1,519 @@
+        current_table_type = base_type
+        for column_token in column_tokens:
+            current_table_schema = self.schema.schema[current_table_type]
+            if column_token in current_table_schema.relations:
+                current_table_type = current_table_schema.relations[column_token]
+                continue
+            elif current_table_schema.attributes_regex.match(column_token):
+                return current_table_type
+        return current_table_type
+
+    def convert_to_1NF_from_unnormalized(self, table_name: str) -> None:
+        """Converts an unnormalized table into first normal form (1NF)
+
+        Removes anticipated repeating groups
+
+        Args:
+            table_name: name of table type to normalize
+        Modifies:
+            database[table_name] transformed to 1NF
+        """
+        table_schema = self.schema.schema[table_name]
+        unnormalized_table = self.database[table_name]
+        repeated_columns = [
+            column
+            for column in unnormalized_table.columns
+            if table_schema.repeating_columns_regex.match(column)
+        ]
+        if repeated_columns == []:
+            return unnormalized_table
+        for column in repeated_columns:
+            base_column_name = column.split("-")[0]
+            if base_column_name in unnormalized_table.columns:
+                # base table has col and col-1. Replace col with col-{n+1}
+                max_repeat = max(
+                    [
+                        int(col.split("-")[-1])
+                        for col in repeated_columns
+                        if col.startswith(base_column_name)
+                    ]
+                )
+                new_column_name = f"{base_column_name}-{max_repeat + 1}"
+                unnormalized_table = unnormalized_table.rename(
+                    columns={base_column_name: new_column_name}
+                )
+                repeated_columns.append(new_column_name)
+        unnormalized_table.loc[:, "temp_id"] = range(0, len(unnormalized_table))
+        static_columns = [
+            column
+            for column in unnormalized_table.columns
+            if column not in repeated_columns
+        ]
+        melted_table = pd.melt(
+            unnormalized_table, id_vars=static_columns, value_vars=repeated_columns
+        )
+        melted_table[["variable", "instance"]] = melted_table["variable"].str.rsplit(
+            "-",
+            n=1,
+            expand=True,  # TODO: split variable
+        )
+        first_normal_form_table = melted_table.pivot(  # noqa: PD010
+            index=static_columns + ["instance"],
+            columns="variable",
+            values="value",
+        ).reset_index()
+        first_normal_form_table = first_normal_form_table.drop(
+            columns=["temp_id", "instance"]
+        )
+        first_normal_form_table = first_normal_form_table[
+            first_normal_form_table["amount"].notna()
```

Comment:
Am I right that this means that "amount" is a required column for all tables undergoing normalization? If so, should this be made a little bit more clear in some way? I almost feel like this step could fit better as a separate step, between 1NF and 3NF, since it feels like custom business logic that doesn't fit in either?